File: HW5 - C++/BoardArray1D.cpp
#include <iostream>
#include "AbstractBoard.h"
#include "BoardArray1D.h"
using namespace std ;

namespace GTU_CSE241_2019_2020{
	

	BoardArray1D::BoardArray1D() : AbstractBoard() , board(nullptr){
		// empty
	}

	BoardArray1D::BoardArray1D( const BoardArray1D &other ) : AbstractBoard(){
		
		setSize( other.row, other.column ) ;

		for (int i = 0; i < row; ++i)
			for (int j = 0; j < column; ++j)
				this->board[ (i * column ) + j ] = other.board[ (i * column ) + j ] ;

	}

	BoardArray1D& BoardArray1D::operator =( const BoardArray1D  &rightSide ){
		
		if( *this == rightSide )
			return *this ;
		setSize( rightSide.row, rightSide.column ) ;

		for (int i = 0; i < row; ++i)
			for (int j = 0; j < column; ++j)
				this->board[ (i * column ) + j ] = rightSide.board[ (i * column ) + j ] ;

		return *this ;

	}
	void BoardArray1D::setSize( int new_row, int new_column ){

		if( new_row < 3 || new_column < 3 ){ // set as default
			
			delete [] board ;

			row = 3 ;
			column = 3 ;

			board = new int [ row * column ] ;
		}
		else if( new_row != row || new_column != column ){

			delete [] board ;

			row = new_row ;
			column = new_column ;

			board = new int [ row * column ] ;
		}
		
		reset() ;	
	}

	
	
	int& BoardArray1D::operator () ( int row_cordinate, int column_cordinate ){

		return board[ ( row_cordinate * column ) + column_cordinate ] ;
	}

	const int& BoardArray1D::operator () ( int row_cordinate, int column_cordinate )const{

		return board[ ( row_cordinate * column ) + column_cordinate ] ;

	}
	
	BoardArray1D::~BoardArray1D(){
		
		delete [] board ;
	}

}

File: HW7 - JAVA/src/HW7/BoardArray1D.java
package HW7;

/**
 * @author murat
 * @since 15.01.2020
 *
 */
/**
 * A class which to represent NPuzzle problem.It keeps data as one dimensional integer array.
 * @see AbstractBoard
 */
public class BoardArray1D extends AbstractBoard {
	/** 
	 * Keep data in one dimensional integer array.
	 */
	private int [] board ;
	/** 
	 * Initialize fields
	 * <p> Calls AbstractBoard() as a  super class constructor
	 * @see AbstractBoard#AbstractBoard()
	 */
	public BoardArray1D(){
		super() ;	
	}
	

	@Override
	public void setSize( int newRow, int newColumn ) {
		
		if( board != null && !( newRow == row && newColumn == column) ) {	

			board = null ;
			System.gc();
		}
		
		if( newRow < 3 || newColumn < 3 ){ // set as default
			
			row = 3 ;
			column = 3 ;
			board = new int [row * column] ;

		}
		else if( newRow != row || newColumn != column ){

			row = newRow ;
			column = newColumn ;
			board = new int [row * column] ;
		}
		
		super.reset() ;
	}

	public int cell(int rowIndex, int columnIndex) {
		
		return board[(rowIndex * column) + columnIndex] ;
	}

	
	public void setCell(int rowIndex, int columnIndex, int setValue) {
		
		board[(rowIndex * column) + columnIndex] = setValue ;
	}
}
